### PR TITLE
Improve any-matcher support and add short-circuit expectations

### PR DESCRIPTION
**Pull request by deltics**
## Summary by CodeRabbit

- **New Features**
  - Added `BeEmpty()`, `BeEmptyOrNil()`, and `BeNil()` matchers supporting diverse Go types for emptiness and nilness checks.
  - Introduced `Require()` to mark expectations as critical, causing immediate test failure on unmet conditions.
  - Enhanced test frame management with a concurrency-safe stack, `With(t)` helper, and support for example tests without `*testing.T`.
  - Provided utilities for temporary variable replacement and restoration in tests.
  - Added `Should()` and `ShouldNot()` matcher methods for non-type-safe matching.
  - Introduced detailed failure reporting options and expanded matcher option support.

- **Bug Fixes**
  - Improved invalid test and error reporting with clearer messages and consistent handling across the framework.
  - Fixed error handling in parallel test setup and panic matching logic.

- **Documentation**
  - Expanded and clarified documentation on test frame setup, matcher usage, type safety distinctions, short-circuit evaluation, and matcher options.
  - Added detailed usage examples for new matchers and test helpers.

- **Refactor**
  - Streamlined matcher interfaces to separate type-safe and any-type matching.
  - Unified invalid test signalling and error handling using centralized functions.
  - Removed legacy nil-check methods and replaced with matcher-based nilness assertions.
  - Simplified internal test frame stack implementation and error handling.
  - Updated test scenario naming and execution flow for clarity.

- **Tests**
  - Added comprehensive tests covering new matchers, test frame stack, nil frame scenarios, example test runner, and requirement enforcement.
  - Updated existing tests to use new test frame setup and matcher methods for improved consistency.
  - Added coverage for new options, error conditions, and utility functions.

- **Chores**
  - Removed obsolete files and internal test frame implementations replaced by new concurrency-safe stack.
  - Improved code comments, formatting, and internal consistency.
  - Corrected minor typos and enhanced option handling in formatting utilities.
  - Added SonarLint configuration for code analysis integration.

### CODE REVIEW

File: map.go [hunk 1]
```diff
@@ -83,19 +86,17 @@ func ContainMap[K comparable, V any](want map[K]V) maps.ContainsMatcher[K, V] {
...
 		)
 	case len(want) == 0:
-		invalidTest(
+		test.Invalid(
 			"ContainMap() called with empty map.",
 			"Did you mean Expect(map).To(EqualMap(<empty map>)) or Expect(map).IsEmpty()?",
 		)
-	default:
-		return maps.ContainsMatcher[K, V]{Expected: want}
 	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply map.go:98]: Explicit returns in the cases would never be reached, making the 100% test coverage goal impossible to attain.

The `test.Invalid` calls are made in cases of a switch with no default; control-flow analysis will be satisfied that the existing return statement will be reached (even though it is not) while ensuring that 100% coverage is attainable.

File: matchers/nilness/isNil.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+		reflect.Map,
+		reflect.Ptr,
+		reflect.Slice:
+		if v := reflect.ValueOf(v); v != reflect.Zero(reflect.TypeOf(v)) && !v.IsNil() {
+			return false
+		}
+		return true
+	}
+	return ErrNotNilable
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/nilness/isNil.go:28]: Good idea, thanks!

File: opt/valueAsString.go [hunk 3]
```diff
@@ -21,9 +21,16 @@ func ValueAsString(v any, opts ...any) string {
...
+
+	spec := "%v"
+	switch {
+	case isString && IsSet(opts, QuotedStrings(false)):
 		spec = "%v"
+	case isString:
+		spec = "%q"
+	case IsSet(opts, AsDeclaration(true)):
+		spec = "%#v"
 	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply opt/valueAsString.go:34]: This is deliberate.  String values are formatted as strings and are subject to the `QuotedStrings` option.  The `AsDeclaration` option is applied only to non-string values.

File: panic.go [hunk 4]
```diff
@@ -6,6 +6,7 @@ import (
 
 	"github.com/blugnu/test/matchers/panics"
 	"github.com/blugnu/test/opt"
+	"github.com/blugnu/test/test"
 )
```
[the remark this answers is not in the document]
Comment (deltics) [reply panic.go:10]: There is no conflict.  Package qualifiers never refer to the current package so a `test` qualification can only refer to the imported package.

While the practice may be confusing in general, in this case the usage is specific and clear enough.

File: parallel.go [hunk 5]
```diff
@@ -1,9 +1,10 @@
 package test
 
 import (
-	"fmt"
 	"reflect"
 	"testing"
+
+	"github.com/blugnu/test/test"
 )
```
[the remark this answers is not in the document]
Comment (deltics) [reply parallel.go:8]: Local (current package) symbols are always unqualified; imported symbols are always qualified.  There is no confusion and no possibility of accidental self-references.

File: runScenarios.go [hunk 6]
```diff
@@ -80,3 +50,27 @@ func RunScenarios[T any](f func(*T, int), scns []T) {
...
+
+	candidate := []string{"Scenario", "scenario", "Name", "name"}
+	for _, c := range candidate {
+		if f := reflect.Indirect(ref).FieldByName(c); f.IsValid() && f.Kind() == reflect.String {
+			return f.String()
+		}
+	}
+
+	return result
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply runScenarios.go:76]: Good catch, thanks!

File: test.go [hunk 7]
```diff
@@ -261,3 +295,21 @@ func runInternal(t *testing.T, f func(*testing.T)) ([]string, []string, TestOutc
...
+	frames := runtime.CallersFrames(pcs)
+
+	frame, more := frames.Next()
+	for more && !strings.HasSuffix(frame.File, "_test.go") {
+		frame, more = frames.Next()
+	}
+
+	_, filename := path.Split(frame.File)
+	return filename
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply test.go:315]: Good catch, thanks!

File: test/nilFrame.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+//
+//	      // this should panic, meeting the expectation established
+//	      // and deferred (above)
+//	      T()
+//	   })
+//	}
+func NilFrame() testframe.Nil {
+	return testframe.Nil{
+		T: T().(testframe.Cleanup),
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply test/nilFrame.go:50]: Good idea, thanks!

File: t.go [hunk 9]
```diff
@@ -31,32 +33,54 @@ var _ TestingT = (*ExampleTestRunner)(nil)
...
+}
+
+func (m ExampleTestRunner) Run(name string, f func(t *testing.T)) bool {
+	return false
 }
+
+func (m *ExampleTestRunner) Fail()                { m.failed = !m.exited && true }
+func (m *ExampleTestRunner) FailNow()             { m.failed = true; m.exited = true }
+func (m *ExampleTestRunner) Failed() bool         { return m.failed }
 func (m ExampleTestRunner) Cleanup(func())        { /* no-op */ }
```
[the remark this answers is not in the document]
Comment (deltics) [reply t.go:84]: The `ExampleTestRunner` is used solely to provide a minimal runner to support runnable examples for pkg docs.  It is not intended to provide a complete implementation.

File: t_test.go [hunk 10]
```diff
@@ -19,42 +23,57 @@ func TestT(t *testing.T) {
...
+	Run("nil", func() {
+		// the panic test must be setup before niling the test frame
+		defer Expect(Panic(testframe.ErrNoTestFrame)).DidOccur()
 
-			With(nil)
-			_ = T()
-		})
+		With(nil)
 	})
+
```
[the remark this answers is not in the document]
Comment (deltics) [reply t_test.go:52]: The `defer` ensures that the `DidOccur()` method is performed *after* the panic has occurred.  `Expect()` is called immediately, `DidOccur()` is deferred.

File: expect.go [hunk 11]
```diff
@@ -99,41 +144,33 @@ func (e expectation[T]) err(msg any) {
...
+	// that will be determined later, if needed)
 	report := e.getTestFailureReporter(opts...)
 	if report == nil {
 		report = matcher
 	}
 
+	// expectation.required may be preset or may have been specified
+	// as an option
+	e.required = e.required || opt.IsSet(opts, opt.IsRequired(true))
+
```
[the remark this answers is not in the document]
Comment (deltics) [reply expect.go:173]: Thanks!

File: matchers/emptiness/matcher.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+	// _type is the type of the subject, if known
+	_type string
+
+	// len is the length of the subject, if known to a maximum of math.MaxInt
+	len int
+
+	// string is the string representation of the length of the subject
+	string
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/emptiness/matcher.go:41]: It is legal to embed fields using `string` as long as the field does not need to be exported (which this does not).  There are no compilation error.

However, the `int` field, `len` is not used and is effectively redundant as written.  So I have chosen instead to re-purpose the `len` field as the string representation of the length (which *is* used), to replace the anonymous `string` field.

File: matchers/emptiness/matcher.go [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+func (m *Matcher) Match(subject any, opts ...any) bool {
+	m.hasLen(subject)
+	switch {
+	case m.isNil:
+		return false
+	case m.hasLength:
+		return m.isZero
+	}
```
[the remark this answers is not in the document]
Comment (deltics) [reply matchers/emptiness/matcher.go:50]: `TreatNilAsEmpty` is applied by the `tryLen` method; if `isNil` is `true`, then the value is `nil` and `TreatEmptyAsNil` was not set.

I have added a clarifying comment explaining this.

File: restore.go [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+	og := *tf.ptr
+	*tf.ptr = r
+
+	// returns a closure over the original variable address and value
+	// that will restore the original value when called
+	//
+	// the (unused) parameter ensures that the returned function is
+	// compatible with the Restore function signature
+	return func(restorable[T]) { *tf.ptr = og }
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply restore.go:23]: Good catch on the `nil` guard; this has been address in the `Original` method rather than here.

Double-replacement is more challengng to resolve and is an edge-case that has side-effects that are likely to be identified and resolved through unexpected test failures in a suite.

An issue will be created for it to be addressed later.

File: test/example.go [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+	"testing"
+
+	"github.com/blugnu/test/internal/testframe"
+)
+
+// Example is used to initialise a testframe for an Example function
+// (that is, a runnable example on pkg.go.dev)
+func Example() {
+	testframe.Push(&example{})
+}
```
[the remark this answers is not in the document]
Comment (deltics) [reply test/example.go:14]: Even deferring popping the frame from the stack would mean that the frame is prematurely popped; the function is intended to push an example runner onto the stack for subsequent example code to use.  The `T()` function returns the most recent frame from the current stack.

I have done some testing and determined that doc examples appear to run in a single goroutine in a given package, so it is correct to say that each example pushes a new frame onto the stack and that those frames are never popped.  However, running doc examples is a short-lived and very specific process and the amount of memory involved is negligible; the complexity that would be necessary to try to manage this more efficiently would be significant and risk introducing more confusing side-effects and bugs than is justified.

For now I have chosen to add a comment explaining the consquences of the current implementation and rationale for not addressing it.

File: runScenarios.go [hunk 16]
```diff
@@ -80,3 +50,30 @@ func RunScenarios[T any](f func(*T, int), scns []T) {
...
+// in the struct and returns its value as a string. If no such field is found,
+// it returns a default name in the format "testcase/000" where 000 is the
+// index of the scenario in the slice.
+func getScenarioName(scn any, idx int) string {
+	result := fmt.Sprintf("testcase-%.3d", idx)
+
+	// if the scenario is not a struct (or pointer to a struct), return
+	// the default result
+	ref := reflect.Indirect(reflect.ValueOf(scn))
+	if ref.Kind() != reflect.Struct {
```
[the remark this answers is not in the document]
Comment (deltics) [reply runScenarios.go:66]: when `scn` is `nil`, `ref.Kind()` returns a value that is not `reflect.Struct`, so the existing implementation works as intended and returns the default name.  I have added a test case to confirm and ensure any change in behaviour in the future is caught.

File: README.md [hunk 17]
```diff
@@ -113,6 +113,61 @@ go get github.com/blugnu/test
...
+The `With(t *testing.T)` function is used to set up a test frame for the current test.  This
+function is typically called at the start of a test function, passing the `*testing.T` value
+from the test function as an argument:
+
+```go
+  func TestDoSomething(t *testing.T) {
+     With(t) // establishes the initial test frame
+     // ...
+  }
+```
```
[the remark this answers is not in the document]
Comment (deltics) [reply README.md]: While the function accepts the `TestingT` interface, in use it will be a `*testing.T` that is supplied since `*testing.T` implements the `TestingT` interface.  The README documents the usage of the function, not the contracted arguments.
